src/libraries/Microsoft.Extensions.Logging/gen/TemplateHelper.cs

```diff
@@ -0,0 +1,346 @@
+                    // classic composite format string
+
+                    pos++;
+                    if (pos == len || (ch = format[pos]) < '0' || ch > '9')
+                    {
+                        // we need an argument index
+                        throw new ArgumentException($"Missing argument index in format string at position {pos}", nameof(format));
+                    }
+
+                    do
+                    {
+                        argIndex = (argIndex * 10) + (ch - '0');
+                        pos++;
+
+                        // make sure we get a suitable end to the argument index
+                        if (pos == len)
+                        {
+                            throw new ArgumentException($"Invalid argument index in format string at position {pos}", nameof(format));
+                        }
+
+                        ch = format[pos];
+                    }
+                    while (ch >= '0' && ch <= '9');
+                }
+                else
+                {
+                    // template-based format string
+
+                    pos++;
+                    if (pos == len)
+                    {
+                        // we need a template name
+                        throw new ArgumentException($"Missing template name in format string at position {pos}", nameof(format));
+                    }
+
+                    ch = format[pos];
+                    if (!ValidTemplateNameChar(ch, true))
+                    {
+                        // we need a template name
+                        throw new ArgumentException($"Missing template name in format string at position {pos}", nameof(format));
+                    }
+
+                    // extract the template name
+                    var start = pos;
+                    do
+                    {
+                        pos++;
+
+                        // make sure we get a suitable end
+                        if (pos == len)
+                        {
+                            throw new ArgumentException($"Invalid template name in format string at position {pos}", nameof(format));
+                        }
+
+                        ch = format[pos];
+                    }
+                    while (ValidTemplateNameChar(ch, false));
+
+                    // get an argument index for the given template
+                    var template = format.Slice(start, pos - start).ToString();
+                    argIndex = templates.IndexOf(template);
+                    if (argIndex < 0)
+                    {
+                        templates.Add(template);
+                        argIndex = numArgs;
+                    }
+                }
+
+                if (argIndex >= numArgs)
+                {
+                    // new max arg count
+                    numArgs = argIndex + 1;
+                }
+
+                // skip whitespace
+                while (pos < len && (ch = format[pos]) == ' ')
```

Comment:
>format [](http://example.com/codeflow?start=42&length=6)

does the format can have different space characters? I mean shoud you use Char.IsWhiteSpace instead?

Reply:
you need to do that in other places to check space char?